src/TestFramework/TestFramework/Assertions/Assert.AreEqual.cs

```diff
@@ -2118,3 +2226,121 @@ private static void ThrowAssertAreNotEqualFailed(object? notExpected, object? ac
         ThrowAssertFailed("Assert.AreNotEqual", finalMessage);
     }
 }
+
+internal static class StringPreviewHelper
+{
+    public static Tuple<string, string, int> CreateStringPreviews(string expected, string actual, int diffIndex, int fullPreviewLength)
+    {
+        int ellipsisLength = 3; // Length of the ellipsis "..."
+
+        if (fullPreviewLength % 2 == 0)
+        {
+            throw new ArgumentException($"{nameof(fullPreviewLength)} must be odd, but it was even.");
+        }
+
+        if (fullPreviewLength < 9)
+        {
+            // Too short context window has additional complication with ellipsis etc.
+            throw new ArgumentException($"{nameof(fullPreviewLength)} cannot be shorter than 9.");
+        }
+
+        // In case we want to instead count runes or text elements we can change it just here.
+        int expectedLength = expected.Length;
+        int actualLength = actual.Length;
+
+        if (diffIndex < 0 || diffIndex > Math.Min(expectedLength, actualLength)) // Not -1 here because the difference can be right after the end of the shorter string.
+        {
+            throw new ArgumentOutOfRangeException(nameof(diffIndex), "diffIndex must be within the bounds of both strings.");
+        }
+
+        int contextLength = (fullPreviewLength - 1) / 2;
+
+        // Diff index must point into the string, the start of the strings will always be shortened the same amount,
+        // because otherwise the diff would happen at the beginning of the string.
+        // So we just care about how far we are from the end of the string, so we can show the maximum amount of info to the user
+        // when diff is really close to the end.
+        string shorterString = expectedLength < actualLength ? expected : actual;
```

Comment:
I wonder if it would be more readable to start with:

```
leftLength = (fullPreviewLength - 1) / 2;
cutStart = Math.Min(0, diffIndex - leftLength);
cutEndShort = Math.Max(cutStart + fullPreviewLength, shorterStringLength);
cutEndLong = Math.Max(cutStart + fullPreviewLength, longerStringLength);
```

This way the string is centered around the diffIndex, if it's long enough, but also handles the scenario of the preview starting at string length or ending at the end of the string.